packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts

```diff
@@ -0,0 +1,228 @@
+    persistedCache: IPersistedCache = new LocalPersistentCache(),
+    logger: ITelemetryBaseLogger,
+    hostSnapshotFetchOptions: ISnapshotOptions | undefined,
+): Promise<boolean> {
+    const odspLogger = createOdspLogger(ChildLogger.create(logger, "PrefetchSnapshot"));
+    const odspResolvedUrl = getOdspResolvedUrl(resolvedUrl);
+    const snapshotUrl = odspResolvedUrl.endpoints.snapshotStorageUrl;
+    let snapshotOptions: ISnapshotOptions = {
+        deltas: 1,
+        channels: 1,
+        blobs: 2,
+    };
+
+    if (hostSnapshotFetchOptions) {
+        snapshotOptions = {...snapshotOptions, ...hostSnapshotFetchOptions};
+    }
+
+    const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+        odspLogger,
+        odspResolvedUrl,
+        getStorageToken,
+        true /* throwOnNullToken */,
+    );
+
+    const snapshotCacheEntry: ICacheEntry = {
+        type: snapshotKey,
+        key: "",
+        file: {
+            resolvedUrl: odspResolvedUrl,
+            docId: odspResolvedUrl.hashedDocumentId,
+        },
+    };
+
+    return getWithRetryForTokenRefresh(async (tokenFetchOptions) => {
+        // No need to ask cache twice - if first request was unsuccessful, cache unlikely to have data on second turn.
+        if (tokenFetchOptions.refresh) {
+            await fetchSnapshot(
+                snapshotUrl,
+                odspResolvedUrl,
+                snapshotOptions,
+                tokenFetchOptions,
+                storageTokenFetcher,
+                odspLogger,
+                persistedCache,
+                snapshotCacheEntry,
+            );
+        } else {
+            await PerformanceEvent.timedExecAsync(
+                odspLogger,
+                { eventName: "ObtainSnapshot" },
+                async (event: PerformanceEvent) => {
+                    const cachedSnapshot: ISnapshotCacheValue | undefined =
+                        await persistedCache.get(snapshotCacheEntry);
+                    const method = cachedSnapshot !== undefined ? "cache" : "network";
+
+                    if (cachedSnapshot === undefined) {
+                        await fetchSnapshot(
+                            snapshotUrl,
+                            odspResolvedUrl,
+                            snapshotOptions,
+                            tokenFetchOptions,
+                            storageTokenFetcher,
+                            odspLogger,
+                            persistedCache,
+                            snapshotCacheEntry,
+                        );
+                    }
+                    event.end({ method });
+                    return cachedSnapshot;
+                },
+            );
+        }
+        if (persistedCache.get(snapshotCacheEntry) !== undefined) {
+            return true;
+        }
+        return false;
+    }).catch((error) => {
+        return false;
+    });
+}
+
+async function fetchSnapshot(
```

Comment:
Is there any way to refactor code and reuse common parts between this function and fetchSnapshotCore()? I worry about future complexity (features similar to redeeming sharing link - we would need to change them in both places, and there is good chance we will forget).

Reply:
I will look that in follow up PR.

Reply:
Let's open an issue for that

Reply:
https://github.com/microsoft/FluidFramework/issues/6018